### PR TITLE
Support SM90 DSA qh16 indexer forward and fix qh32 sparse backward

### LINKED ISSUE #373: Can SM90 DSA indexer_forward support qhead_per_kv_head=16
Hi, I have a question about the DSA cuDNN frontend implementation.

In cudnn-frontend/python/cudnn/deepseek_sparse_attention/indexer_forward/_interface_sm90.py
the SM90 indexer_fwd path currently defines:

_SUPPORTED_QHPKV = (32, 64)
So qhead_per_kv_head=16 is rejected.

I’d like to ask whether the underlying operator implementation can support qhead_per_kv_head=16.

### LINKED ISSUE #385: [Bug] Sparse Attention Backward (`_FLASH_ATTN_BWD_SM90`) produces catastrophically wrong gradients on SM90
## [Bug] Sparse Attention Backward (`_FLASH_ATTN_BWD_SM90`) produces catastrophically wrong gradients on SM90

**Describe the bug**

The `_FLASH_ATTN_BWD_SM90` CuTe DSL kernel in the sparse attention backward pass produces **directionally incorrect gradients** with catastrophic magnitude (~1e11 vs correct ~1e2). This is **not a precision or scaling issue** — the output has near-zero cosine similarity with the correct reference, indicating a fundamental algorithmic bug in the compiled kernel.

Symptom summary:

| Metric | Correct (fp64/tilelang) | cuDNN Output |
|--------|------------------------|--------------|
| dq absmax | 15.3 | **6.79e11** |
| dkv absmax | 166 | **1.26e11** |
| Cosine similarity | 1.0 | **~0.0** |
| Max relative error | ~0.3% | **~1e10** |

Anomaly rate (elements with >1% relative error):
- **dq**: 49% of all elements (global corruption)
- **dkv**: 11% of elements, but 10% have >1000% error (concentrated in specific KV rows)

We isolated the bug with a 4-layer comparison:
1. **fp64 reference** (hand-written FlashAttention backward): dkv = 166 ✓
2. **tilelang kernel** (independent implementation): dkv = 166, cosine 0.9999 vs fp64 ✓
3. **cuDNN wrapper** (`sparse_mla_bwd_cudnn`): dkv = 1.26e11, cosine ~0 ✗
4. **Direct `_FLASH_ATTN_BWD_SM90` call** (bypassing all Python preprocessing): inputs all clean and normal (q~9, kv~14, do~5.6, lse~15, indices all ≤7248, no NaN/Inf), but output dkv = 1.26e11, cosine ~0 ✗

This confirms the bug is **inside the compiled CuTe kernel**, not in any Python wrapper / index conversion / LSE handling.

We also tested all version combinations — the broken output is **bit-identical** across them, so this is not a version-compatibility issue:

| cudnn_frontend | cudnn_backend (cu13) | CUTLASS | Result |
|----------------|----------------------|---------|--------|
| 1.25.0 | 9.19.0.56 | 4.5.0 | dkv = 1.26e11 (bug) |
| 1.26.0 | 9.19.0.56 | 4.5.0 | dkv = 1.26e11 (bit-identical) |
| 1.26.0 | 9.24.0.43 | 4.5.0 | dkv = 1.26e11 (bit-identical) |
| 1.25/1.26 | — | 4.6.0 | API incompatible (`map_dataclass_to_tuple`) |

**Expected behavior**

The backward kernel should produce gradients matching the sparse FlashAttention backward formula:
```
Δ  = rowsum(O ⊙ dO)
P  = exp(Q @ K^T · scale − LSE)
dP = dO @ V^T
dS = P ⊙ (dP − Δ) · scale
dQ = dS @ K
dK = dS^T @ Q
dV = P^T @ dO
```
Our fp64 hand-written implementation and the tilelang kernel both produce dq≈15.3 / dkv≈166 and agree with each other (cosine 1.0000). Every intermediate is normal (|P|≤1, |dP|~35, |Δ|~11, |dS|~0.6) — the formula cannot produce 1e11. cuDNN should produce the same, but instead outputs ~1e11.

**System Environment (please complete the following information):**
 - cudnn_frontend version: v1.25.0 and v1.26.0 (both affected, bit-identical broken output)
 - cudnn_backend version: nvidia-cudnn-cu13 9.19.0.56 and 9.24.0.43 (no impact — DSA backward is pure CuTe DSL, does not call the C++ backend)
 - GPU arch: NVIDIA H200 (SM90)
 - cuda runtime version: 13.3
 - cuda driver version: (available on request)
 - host compiler: (available on request)
 - OS: Linux 5.15.0-25-generic
 - nvidia-cutlass-dsl: 4.5.0 (required by cudnn_frontend 1.26.0)
 - PyTorch: 2.8.0+cu133

**API logs**

Not yet collected. The bug is deterministic and fully reproducible with our minimal test case; we can attach cudnn_frontend / cudnn_backend logs if needed.
```
// For cudnn_frontend
export CUDNN_FRONTEND_LOG_FILE=fe.log
export CUDNN_FRONTEND_LOG_INFO=1

// For cudnn_backend
export CUDNN_LOGLEVEL_DBG=3
export CUDNN_LOGDEST_DBG=be.log
```

**To Reproduce**

1. Load a real training dump of the main-kernel inputs (`q`, `kv`, `do`, `lse`, `topk_idxs`, `dpsum`), all with normal magnitude and no NaN/Inf.
2. Call `_FLASH_ATTN_BWD_SM90(q, kv, out=o, dout=do, lse, topk_idxs, softmax_scale, ...)` directly.
3. Observe the output: `dq` absmax ≈ 6.79e11, `dkv` absmax ≈ 1.26e11 (correct values are 15.3 / 166), with ~0 cosine similarity vs the fp64/tilelang reference.

Observed output on our setup:
```
=== Comparison (baseline = fp64) ===
  dq  tilelang vs fp64   cosine=0.999998  max_rel_err=3.054e-03
  dq  cuDNN    vs fp64   cosine=0.000000  max_rel_err=4.428e+10
  dkv tilelang vs fp64   cosine=0.999997  max_rel_err=3.114e-03
  dkv cuDNN    vs fp64   cosine=-0.001002  max_rel_err=7.574e+08

Direct kernel call (inputs all clean):
  q     : absmax=9.3750e+00 nan=0 inf=0
  kv    : absmax=1.4000e+01 nan=0 inf=0
  o     : absmax=1.5723e-01 nan=0 inf=0
  do    : absmax=5.5938e+00 nan=0 inf=0
  lse   : absmax=1.5411e+01 nan=0 inf=0
  topk_global: absmax=9.0500e+02 nan=0 inf=0
  → dq_kernel : absmax=6.7860e+11   dkv_kernel: absmax=1.2563e+11
  → dq_kernel  vs fp64  cosine=0.000000
  → dkv_kernel vs fp64  cosine=-0.001002

Anomaly rate (cuDNN vs fp64):
  dq:   49.11% elements with >1% relative error
  dkv:  11.28% elements with >1% relative error (9.76% with >1000%)
```

**Additional context**

Code location — the bug is in the compiled kernel invoked at:
```
cudnn/deepseek_sparse_attention/sparse_attention_backward/_interface_sm90.py:295
    flash_attn_bwd_sm90.compile_cache[compile_key](...)
```
which compiles the CuTe DSL kernel defined in:
```
cudnn/deepseek_sparse_attention/sparse_attention_backward/dsa_bwd_sm90.py
cudnn/deepseek_sparse_attention/utils/sm90/*
```

Impact: this makes the cuDNN sparse attention backward unusable for training — gradients accumulate across layers, overflow bf16 max (3.39e38), and become NaN. 

The dumped kernel inputs are available here: https://drive.google.com/file/d/1tE3KhDwmstssyovtaHcTvtDJ5xvCiy6z/view?usp=drive_link

### PR DESCRIPTION
**Pull request by jiayus-nvidia**
## Summary

  - Add SM90 DSA indexer-forward support for `qhead_per_kv_head=16`.
  - Fix catastrophic SM90 sparse-attention backward errors for `qhead_per_kv_head=32`.
  - Handle qh32 directly inside the main kernel without introducing padding or copy kernels.
  - Update DSA documentation and regression coverage.

  ## Details

  ### SM90 indexer forward with qh16

  A qh16 CTA packs eight query tokens. Their ratio-causal limits can cross a 64-column boundary, so masking only the rightmost N tile is insufficient.

  This change:

  - Enables qh16 in the SM90 interface.
  - Applies elementwise causal masking to both rightmost N tiles for qh16.
  - Adds coverage for causal block boundaries and THD variable-length tails.

  ### SM90 sparse-attention backward

  The backward kernel previously flattened query heads and query positions into a packed-M dimension. With qh32, one 64-row CTA covered two queries
  while loading only the first query's top-k row. The second query therefore combined its Q/LSE/dPsum values with the wrong KV indices, producing
  catastrophically large gradients.

  This change:

  - Separates query position from the query-head tile in the kernel layout.
  - Schedules each CTA for exactly one query and one head tile.
  - Keeps the physical 64-row WGMMA tile and lets TMA zero-fill unused qh32 rows.
  - Neutralizes invalid rows with `LSE=+inf` and `dPsum=0`.
  - Uses the current query's top-k row in both warp groups.
  - Drops invalid dQ rows through TMA out-of-bounds predication.
  - Keeps `qhead_per_kv_head` in the main-kernel compile key.
  - Removes the Python padding fallback and its additional pad/copy kernels.

  The DSA reference was also updated for the MLA shape where QK uses 576 dimensions and V uses the first 512 dimensions.

  ## Testing

  Validated on an NVIDIA H100 (SM90):

  - SM90 qh16 indexer-forward regressions: `2 passed`
  - Sparse-attention backward qh32/qh64 matrix, with and without `topk_length`: `5 passed`
  - Verified qh32, qh64, and qh128 against the PyTorch reference.
  - Verified head-major `out`/`dout` layouts from the original reproduction.
  - Verified that the qh32 path does not call `torch.nn.functional.pad`.

  Using the original #385 kernel dump:

  - `dq absmax`: approximately `6.786e11` → `15.3125`
  - dKV FP32 accumulator absmax: approximately `1.256e11` → `165.75`

  The corrected values match the reported reference magnitudes.

  Closes #373
  Fixes #385

## Summary by CodeRabbit

- **New Features**
  - Enabled additional SM90 Indexer Forward configurations with `qhead_per_kv_head=16`.
  - Improved support for MLA-style value tensors using 576-wide dimensions.
- **Bug Fixes**
  - Updated causal-masking at query-to-key block boundaries for the new `qh16` behavior.
  - Strengthened masking correctness for variable-length, multi-batch runs, including tail key positions.
- **Documentation**
  - Clarified DSA Indexer Forward constraints and hardware support (e.g., `head_dim=128`, SM90 vs SM100, and `H_kv` requirements).
- **Tests**
  - Added SM90 Indexer Forward and sparse-attention backward coverage, including `qh32` without padding.

### CODE REVIEW

**Comment by Anerudhan** [comment]
@cudnn-ci-bot run

File: docs/fe-oss-apis/dsa.md
Comment (saltyminty) [inline docs/fe-oss-apis/dsa.md]: NIT: "Indexer Forward and Indexer Top-K remain SM100+ only." is now outdated since this PR adds SM90 support

**Review by saltyminty — APPROVED** [review]
